pkg/server/plugin/zmq/broker.go

```diff
@@ -50,81 +50,96 @@ const (
 	Shutdown = "\003"
 )
 
-// Broker implements the Paranoid Pirate queue described in the zguide:
-// http://zguide.zeromq.org/py:all#Robust-Reliable-Queuing-Paranoid-Pirate-Pattern
+// parcel is used to multiplex the chan with zmq worker
+type parcel struct {
+	respChan chan []byte
+	frame    []byte
+	retry    int
+}
+
+func newParcel(frame []byte) *parcel {
+	return &parcel{
+		respChan: make(chan []byte),
+		frame:    frame,
+		retry:    0,
+	}
+}
+
+// Broker implements the Paranoid Pirate queue described as follow:
 // Related RFC: https://rfc.zeromq.org/spec:6/PPP
+// refs: http://zguide.zeromq.org/py:all#Robust-Reliable-Queuing-Paranoid-Pirate-Pattern
 // with the addition of:
 //
 // 1. Shutdown signal, which signifies a normal termination of worker to provide
 //    a fast path of worker removal
 type Broker struct {
-	name string
-	// NOTE: goroutines are caller of plugin, so frontend is Go side,
-	// backend is plugin side
-	frontend      *goczmq.Sock
-	backend       *goczmq.Sock
-	bothPoller    *goczmq.Poller
-	backendPoller *goczmq.Poller
-	workers       workerQueue
-	freshWorkers  chan []byte
-	logger        *logrus.Entry
+	name        string
+	backendAddr string
+	frontend    chan [][]byte
+	recvChan    chan *parcel
+	addressChan map[string]chan []byte
+	timeout     chan string
+	workers     workerQueue
+	logger      *logrus.Entry
+	stop        chan int
```

Comment:
if you have still got time: there are two components to this broker: one is a broker and the other is the channeler. There are some struct members that are only used in one component but not the other, perhaps we could make the distinction more obvious by creating separate structs for these two components.